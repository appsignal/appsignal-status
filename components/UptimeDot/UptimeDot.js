import PropTypes from "prop-types";
import Tippy from "@tippyjs/react";
import dayjs from "dayjs";
import advancedFormat from "dayjs/plugin/advancedFormat";
dayjs.extend(advancedFormat);

export const formatRegion = (region) => {
  return region
    .split(/-| /)
    .map((region) => region.charAt(0).toUpperCase() + region.slice(1))
    .join(" ");
};

const state = (timeserieValues) =>
  Object.values(timeserieValues).reduce((a, b) => a + b) > 0 ? "down" : "up";

export const downtimeSummary = (timeserieValues) => {
  if (state(timeserieValues) === "up") return <p>No outage</p>;

  return Object.keys(timeserieValues).map((region) => {
    const downtime = timeserieValues[region];
    if (downtime > 0) {
      return (
        <p key={region}>
          {formatRegion(region)} down for {downtime} minutes
        </p>
      );
    }
  });
};

const UptimeDot = ({ timeserie }) => {
  const uptimeDate = dayjs(timeserie.timestamp).format("MMM. Do");

  return (
    <Tippy
      content={
        <div className="text-center text-sm">
          <p className="text-gray-200">{uptimeDate}</p>
          <div>{downtimeSummary(timeserie.values)}</div>
        </div>
      }
    >
      <div
        className={`h-8 flex-grow ${
          state(timeserie.values) === "up" ? "bg-green-500" : "bg-red-500"
        } rounded`}
      />
    </Tippy>
  );
};

UptimeDot.propTypes = {
  timeserie: PropTypes.shape({
    timestamp: PropTypes.string.isRequired,
    values: PropTypes.object.isRequired,
  }),
};

export default UptimeDot;
