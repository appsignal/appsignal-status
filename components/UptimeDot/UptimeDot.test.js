import { render, screen, fireEvent } from "@testing-library/react";

import UptimeDot, { downtimeSummary } from "./UptimeDot";

const up = {
  timestamp: "2021-08-06T22:00:00.000Z",
  values: {
    europe: 0,
    "north-america": 0,
    "south-america": 0,
    "asia-pacific": 0,
  },
};

const down = {
  timestamp: "2021-08-06T22:00:00.000Z",
  values: {
    europe: 60,
    "north-america": 60,
    "south-america": 60,
    "asia-pacific": 60,
  },
};

const build = (props = {}) => {
  return render(<UptimeDot timeserie={up} threshold={5} {...props} />);
};

describe("UptimeDot", () => {
  test("renders without problems", () => {
    const { container } = build();
    expect(container).toMatchSnapshot();
  });

  test("dot is green when up", () => {
    const { container } = build({ timeserie: up });
    const dot = container.querySelector("div");
    expect(dot.classList).toContain("bg-green-500");
  });

  test("dot is red when down and above threshold", () => {
    const { container } = build({ timeserie: down, threshold: 0 });
    const dot = container.querySelector("div");
    expect(dot.classList).toContain("bg-red-500");
  });

  test("dot is green when down and under threshold", () => {
    const { container } = build({ timeserie: down, threshold: 80 });
    const dot = container.querySelector("div");
    expect(dot.classList).toContain("bg-green-500");
  });

  test("renders a tooltip when hovering", async () => {
    const { container } = build({ timeserie: up, threshold: 5 });
    const dot = container.querySelector("div");

    fireEvent.mouseEnter(dot);

    expect(await screen.findByText("Aug. 6th")).toBeInTheDocument();
    expect(
      screen.getByText("No downtimes under 5 minutes")
    ).toBeInTheDocument();
  });

  test("threshold defaults to 5 when undefined", async () => {
    const { container } = build({ timeserie: up, threshold: undefined });
    const dot = container.querySelector("div");

    fireEvent.mouseEnter(dot);

    expect(await screen.findByText("Aug. 6th")).toBeInTheDocument();
    expect(
      screen.getByText("No downtimes under 5 minutes")
    ).toBeInTheDocument();
  });
});

describe("#downtimeSummary", () => {
  test("returns downtime for different regions", () => {
    render(downtimeSummary(down, 5));
    expect(screen.getByText("Europe down for 60 minutes")).toBeInTheDocument();
    expect(
      screen.getByText("North America down for 60 minutes")
    ).toBeInTheDocument();
    expect(
      screen.getByText("South America down for 60 minutes")
    ).toBeInTheDocument();
    expect(
      screen.getByText("Asia Pacific down for 60 minutes")
    ).toBeInTheDocument();
  });

  test("returns 'no downtimes under N minutes' if nothing was down", () => {
    render(downtimeSummary(up, 10));
    expect(
      screen.getByText("No downtimes under 10 minutes")
    ).toBeInTheDocument();
  });
});
