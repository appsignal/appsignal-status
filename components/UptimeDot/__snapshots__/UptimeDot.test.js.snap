// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UptimeDot renders without problems 1`] = `
<div>
  <div
    class="h-8 flex-grow bg-green-500 rounded"
    data-testid="uptimeDot"
  />
</div>
`;
